# 0019-ruby-version.patch
#
# DE DONDE SALE. Parche 0013 de los 37 de syntax-transform/3.1/ del PR https://github.com/mkxp-z/mkxp-z/pull/304
# de mkxp-z, de white-axe (rama syntax-transform, 0be32ad119b054749cf3db8c72434f8fa1ac83ea).
# Ensenan al parser y al nucleo de Ruby 3.1 la gramatica y la semantica de
# Ruby 1.8, que es en lo que estan escritos cinco de los ocho juegos del censo.
#
# COMO SE APLICA. Sobre src/ruby-3.1.7 DESPUES de los parches 0001-0005 y en
# orden de numero: el 0006 (prelude) monta el mecanismo que los otros 36 usan,
# y varios de ellos tocan el mismo fichero, asi que el orden no es decorativo.
#
#     cd src/ruby-3.1.7
#     sed -i 's/\r$//' string.c
#     for n in $(seq 6 42); do
#       patch -p1 -i "../../patches/$(printf '%04d' $n)-"*.patch || break
#     done
#     rm -f parse.c id.h insns.inc insns_info.inc opt_sc.inc optinsn.inc \
#           optunifs.inc vm.inc vmtc.inc mjit_compile.inc
#
# Ese `rm` NO es opcional y cuesta una tarde descubrirlo. El tarball de Ruby
# trae parse.c y los .inc YA GENERADOS en el propio directorio de fuentes, y
# un `#include "insns.inc"` desde compile.c mira primero la carpeta del fichero
# que incluye: aunque el build regenere los suyos al lado del Makefile, gana el
# viejo. Con parse.c viejo los ocho parches de gramatica no llegan al binario y
# NADA lo dice; con los .inc viejos, al menos, no compila
# ('YARVINSN_mkxp_syntax_transform_raise_invalid_break' undeclared).
#
# El cuerpo del diff, de aqui abajo, es el del PR VERBATIM: los 37 aplicaron
# sin fuzz y sin rechazos sobre este arbol, asi que no habia nada que adaptar.
# Lo unico que hay que vigilar es string.c, que en este repositorio esta en
# CRLF y al que `patch` le deja el fichero entero en LF; se normaliza antes,
# para que ese cambio sea deliberado y no un efecto secundario.
#
# COMPROBADO. Con los 37 puestos, bison 3.8.2 acepta la gramatica (parse.y
# declara %expect 0: cualquier conflicto nuevo habria sido un error, y no hubo
# ninguno) y CRuby 3.1.7 cruza-compila para aarch64 con devkitA64 16.1.0 hasta
# libruby-static.a (44,7 MB), con los simbolos del mecanismo dentro
# (mkxp_syntax_transform_set_next_eval, mkxp_ec_is_syntax_transform_active,
# mkxp_syntax_transform_target_ruby_version_*) y 14 instrucciones nuevas en
# insns.inc. No se ha ejecutado: eso pide consola.
#
# QUE HACE FALTA ADEMAS. Estos parches por si solos NO cambian el
# comportamiento: mkxp_syntax_transform_target_ruby_version_* nace en -1 y la
# transformacion queda apagada, asi que el interprete se comporta como el de
# hoy. Encenderla es cosa del lado de mkxp-z (armar
# mkxp_syntax_transform_set_next_eval(1) alrededor del eval de cada seccion de
# Scripts.rxdata y fijar la version objetivo desde mkxp.json), que en este
# repositorio todavia no esta hecho.
#
# Licencia: GPL-2.0-or-later, la del PR. La cabecera original va debajo.
# ---------------------------------------------------------------------------

#
# This file is part of mkxp.
#
# Copyright (C) 2026 The mkxp-z authors
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# This patch changes the values of the toplevel constants `RUBY_VERSION` and
# `RUBY_ENGINE_VERSION` to be consistent with the syntax transform target Ruby
# version in execution contexts where syntax transform is active. The values
# of the constants remain unchanged in execution contexts where syntax transform
# is not active.

--- a/eval.c
+++ b/eval.c
@@ -428,6 +428,18 @@ rb_mod_s_constants(int argc, VALUE *argv, VALUE mod)
             }
         }
 
+        /* Make sure the constants `RUBY_VERSION` and `RUBY_ENGINE_VERSION` are listed in the output of `Module::constants` */
+        {
+            VALUE symbol = RB_ID2SYM(rb_intern("RUBY_VERSION"));
+            if (!rb_ary_includes(ary, symbol)) {
+                rb_ary_push(ary, symbol);
+            }
+            symbol = RB_ID2SYM(rb_intern("RUBY_ENGINE_VERSION"));
+            if (!rb_ary_includes(ary, symbol)) {
+                rb_ary_push(ary, symbol);
+            }
+        }
+
         /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Module::constants` from an array of symbols to an array of strings */
         if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
             long length = RARRAY_LEN(ary);
--- a/variable.c
+++ b/variable.c
@@ -2684,6 +2684,21 @@ rb_const_get_0(VALUE klass, ID id, int exclude, int recurse, int visibility)
             return Qfalse;
         }
     }
+    /* Make sure the constants `RUBY_VERSION` and `RUBY_ENGINE_VERSION` are defined according to the syntax transform target */
+    if (id == rb_intern("RUBY_VERSION") || id == rb_intern("RUBY_ENGINE_VERSION")) {
+        extern unsigned int mkxp_syntax_transform_target_ruby_version_major, mkxp_syntax_transform_target_ruby_version_minor, mkxp_syntax_transform_target_ruby_version_teeny;
+        if (mkxp_ec_is_syntax_transform_active(-1, -1, -1) && mkxp_syntax_transform_target_ruby_version_major != (unsigned int)-1 && mkxp_syntax_transform_target_ruby_version_minor != (unsigned int)-1 && mkxp_syntax_transform_target_ruby_version_teeny != (unsigned int)-1) {
+            VALUE ary = rb_ary_new_capa(3);
+            rb_ary_push(ary, RB_INT2FIX(mkxp_syntax_transform_target_ruby_version_major));
+            rb_ary_push(ary, RB_INT2FIX(mkxp_syntax_transform_target_ruby_version_minor));
+            rb_ary_push(ary, RB_INT2FIX(mkxp_syntax_transform_target_ruby_version_teeny));
+            return rb_ary_join(ary, rb_str_new_cstr("."));
+        }
+        else {
+            extern char ruby_version[];
+            return rb_str_new_cstr(ruby_version);
+        }
+    }
     return rb_const_missing(klass, ID2SYM(id));
 }
 
--- a/version.c
+++ b/version.c
@@ -57,7 +57,7 @@ Init_version(void)
     /*
      * The running version of ruby
      */
-    rb_define_global_const("RUBY_VERSION", (version = MKSTR(version)));
+    /*rb_define_global_const("RUBY_VERSION", (version = MKSTR(version)));*/
     /*
      * The date this ruby was released
      */
@@ -87,7 +87,7 @@ Init_version(void)
     /*
      * The version of the engine or interpreter this ruby uses.
      */
-    rb_define_global_const("RUBY_ENGINE_VERSION", (1 ? version : MKSTR(version)));
+    /*rb_define_global_const("RUBY_ENGINE_VERSION", (1 ? version : MKSTR(version)));*/
 
     rb_provide("ruby2_keywords.rb");
 }
